#160: Extract the filter/paging/sorting from the http request

SchlottiP opened the pull request:
resolves: https://github.com/pace/bricks/issues/160

threez commented on http/jsonapi/runtime/standart_params.go:
```diff
@@ -0,0 +1,117 @@
+		} else {
+			query.Offset(pageSize * (pageNr - 1))
+		}
+		query.Limit(pageSize)
```
would be good if we can set a min,max range for the pageSize

threez commented on http/jsonapi/runtime/standart_params.go:
```diff
@@ -0,0 +1,117 @@
+				order = " DESC"
+			}
+			val = strings.TrimPrefix(val, "-")
+			query.Order(mapper(val) + order)
```
I think we need to provide an array of allowed sorting keys, otherwise we will allow SQLi (https://en.wikipedia.org/wiki/SQL_injection)

threez commented on http/jsonapi/runtime/standart_params.go:
```diff
@@ -0,0 +1,117 @@
+				filterValues = append(filterValues, strings.Split(val, ",")...)
+			}
+			if len(filterValues) == 1 {
+				query.Where(name+"=?", filterValues[0])
```
the same issue with allowed filter values here, to prevent SQLi

threez commented on http/jsonapi/runtime/standart_params.go:
```diff
@@ -0,0 +1,117 @@
+				query.Where(name+"=?", filterValues[0])
+				continue
+			}
+			query.Where(name+" IN (?)", strings.Join(filterValues, ","))
```
and here (SQLi)

threez commented:
@SchlottiP generally looks good but we should add some protection to the query generator

SchlottiP commented:
* added possibility to mark a sort or filter parameter as invalid (with the mapper)
* Added max. and min. for the page size 
* Did not anything to prevent sql injection, because ORMs should do this by preparing the statement bevor replacing the `?` with values, (go-pg is SQL injection save https://github.com/go-pg/pg/wiki/Writing-Queries)

threez commented on http/jsonapi/runtime/standart_params.go:
```diff
@@ -4,24 +4,46 @@
 
 // Query based on orm.Query but only the needed function
 // guaranties testability of all of the following functions
+// An Implementation MUST prepare statements to prevent SQL injection
```
We can't use prepared statements as we are using PGBouncer with transaction mase isolation, is this really required?

threez commented on http/jsonapi/runtime/standart_params.go:
```diff
@@ -4,24 +4,46 @@
+func init() {
+	err := env.Parse(&cfg)
+	if err != nil {
+		log.Fatalf("Failed to parse server environment: %v", err)
```
lets be more specific, otherwise we don't understand right from the message, the cause of the issue:

```suggestion
		log.Fatalf("Failed to parse jsonapi params from environment: %v", err)
```

threez commented on http/jsonapi/runtime/standart_params.go:
```diff
@@ -52,47 +78,73 @@ func PagingFromRequest(r *http.Request) (QueryOption, error) {
+// to a database column name.
+// Will always return a QueryOptions function with all valid filters (can be a none)
+// if any filter are invalid a error with a list of all invalid filters is returned
+// Is SQLi save if the query uses prepared statements
```
same on the prepared statements

threez commented on http/jsonapi/runtime/standart_params.go:
```diff
@@ -102,16 +154,21 @@ func FilterFromRequest(r *http.Request, mapper Mapper) (QueryOption, error) {
 // DefaultMapper maps an query field directly name to the same name as database column name
-func DefaultMapper(in string) string {
-	return in
+func DefaultMapper(in string) (string, bool) {
```
so should we have a mapper now, that is getting a list and returns true if something is in the list? maybe even a map ...

SchlottiP replied:
a Map only allows a explicit whitelist of parameters
a function allows something like a regex to check if the fieldname is valid
=> Both an option, a map is more restrictive by default. I will change it to use a map

threez commented:
@SchlottiP please rebase on the current master so that the amout of changes is reduced

SchlottiP commented:
rebased Branch, kann you review @threez or should I ask Marius or any other person of the team?

threez commented on http/jsonapi/runtime/standart_params.go:
```diff
@@ -0,0 +1,196 @@
+			}
+
+			if len(filterValues) == 1 {
+				query.Where("?=?", name, filterValues[0])
```
not sure it that works, could you add integration tests in general

SchlottiP replied:
Still not Sure if this was okay, changed it before the test was ready

threez commented on Makefile:
```diff
@@ -46,3 +46,6 @@ testserver:
 ci:
 	go test -mod=vendor -count=1 -v -cover -race -covermode=atomic -coverprofile=coverage.out ./...
+
+testone:
```
this may be a left over? rest looks good
